flink-core/src/test/java/org/apache/flink/core/fs/AbstractCloseableRegistryTest.java

```diff
@@ -0,0 +1,230 @@
+		for (Thread t : streamOpenThreads) {
+			t.join();
+		}
+	}
+
+	@Test
+	public void testClose() throws Exception {
+
+		setup();
+		startThreads(Integer.MAX_VALUE);
+
+		for (int i = 0; i < 5; ++i) {
+			System.gc();
+			Thread.sleep(40);
+		}
+
+		closeableRegistry.close();
+
+		joinThreads();
+
+		Assert.assertEquals(0, unclosedCounter.get());
+		Assert.assertEquals(0, closeableRegistry.getNumberOfRegisteredCloseables());
+
+		final C testCloseable = spy(createCloseable());
+
+		try {
+
+			closeableRegistry.registerClosable(testCloseable);
+
+			Assert.fail("Closed registry should not accept closeables!");
+
+		} catch (IOException expected) {
+			//expected
+		}
+
+		Assert.assertEquals(0, unclosedCounter.get());
+		Assert.assertEquals(0, closeableRegistry.getNumberOfRegisteredCloseables());
+		verify(testCloseable).close();
+	}
+
+	@Test
+	public void testNonBlockingClose() throws Exception {
+		setup();
+
+		final OneShotLatch waitRegistryClosedLatch = new OneShotLatch();
+		final OneShotLatch blockCloseLatch = new OneShotLatch();
+
+		final C spyCloseable = spy(createCloseable());
+
+		doAnswer(invocationOnMock -> {
+			invocationOnMock.callRealMethod();
+			waitRegistryClosedLatch.trigger();
+			blockCloseLatch.await();
+			return null;
+		}).when(spyCloseable).close();
+
+		closeableRegistry.registerClosable(spyCloseable);
+
+		Assert.assertEquals(1, closeableRegistry.getNumberOfRegisteredCloseables());
+
+		Thread closer = new Thread(() -> {
+			try {
+				closeableRegistry.close();
+			} catch (IOException ignore) {
+
+			}
+		});
+
+		closer.start();
+		waitRegistryClosedLatch.await();
+
+		final C testCloseable = spy(createCloseable());
+
+		try {
+			closeableRegistry.registerClosable(testCloseable);
+			Assert.fail("Closed registry should not accept closeables!");
+		}catch (IOException ignore) {
+		}
+
+		blockCloseLatch.trigger();
+		closer.join();
+
+		verify(spyCloseable).close();
+		verify(testCloseable).close();
+		Assert.assertEquals(0, closeableRegistry.getNumberOfRegisteredCloseables());
+	}
+
+	protected static abstract class ProducerThread<C extends Closeable, T> extends Thread {
+
+		protected final AbstractCloseableRegistry<C, T> registry;
+		protected final AtomicInteger refCount;
+		protected int maxStreams;
+
+		public ProducerThread(AbstractCloseableRegistry<C, T> registry, AtomicInteger refCount, int maxStreams) {
+			this.registry = registry;
+			this.refCount = refCount;
+			this.maxStreams = maxStreams;
+		}
+
+		public int getMaxStreams() {
+			return maxStreams;
+		}
+
+		public void setMaxStreams(int maxStreams) {
+			this.maxStreams = maxStreams;
+		}
+
+		protected abstract void createAndRegisterStream() throws IOException;
+
+		@Override
+		public void run() {
+			try {
+				while (maxStreams > 0) {
+
+					createAndRegisterStream();
+
+					try {
+						Thread.sleep(2);
+					} catch (InterruptedException ignored) {}
+
+					if (maxStreams != Integer.MAX_VALUE) {
+						--maxStreams;
```

Comment:
nit: I always find it strange when a field that should be a final is also used as a "counter". In those cases there should probably be a field `numStream` and you loop until `maxStreams == openStreams`.